### PR TITLE
Simplify RecoverableText

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs [hunk 1]
```diff
@@ -15,129 +16,117 @@ namespace Microsoft.CodeAnalysis
     /// <summary>
     /// A recoverable TextAndVersion source that saves its text to temporary storage.
     /// </summary>
-    internal class RecoverableTextAndVersion : ValueSource<TextAndVersion>, ITextVersionable
+    internal sealed class RecoverableTextAndVersion : ValueSource<TextAndVersion>, ITextVersionable
     {
-        private readonly ITemporaryStorageServiceInternal _storageService;
+        private readonly HostWorkspaceServices _services;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:21]: possible to relax to SolutionServices?

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs [hunk 2]
```diff
@@ -15,129 +16,117 @@ namespace Microsoft.CodeAnalysis
...
-                value = TextAndVersion.Create(text, _version, _filePath, _loadDiagnostic);
-                return true;
+                if (recoverableText.TryGetValue(out var text))
+                {
+                    value = TextAndVersion.Create(text, recoverableText.Version, filePath: null, recoverableText.LoadDiagnostic);
+                    return true;
+                }
             }
-            else
+            else if (((ValueSource<TextAndVersion>)sourceOrRecoverableText).TryGetValue(out value))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:47]:
```suggestion
            else
            {
                return ((ValueSource<TextAndVersion>)sourceOrRecoverableText).TryGetValue(out value);
            }
```

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs [hunk 3]
```diff
@@ -15,129 +16,117 @@ namespace Microsoft.CodeAnalysis
...
-                }
+                // replace initial source with recovarable text, if it hasn't been replaced already:
+                Interlocked.CompareExchange(
+                    ref _initialSourceOrRecoverableText,
+                    value: new RecoverableText(source.GetValue(cancellationToken), _services),
+                    comparand: source);
             }
 
-            return TextAndVersion.Create(_text.GetValue(cancellationToken), _version, _filePath, _loadDiagnostic);
+            var recoverableText = (RecoverableText)_initialSourceOrRecoverableText;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:85]: safe to just read this?  do you need some sort of barrier here?

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs [hunk 4]
```diff
@@ -15,129 +16,117 @@ namespace Microsoft.CodeAnalysis
...
     /// </summary>
-    internal class RecoverableTextAndVersion : ValueSource<TextAndVersion>, ITextVersionable
+    internal sealed class RecoverableTextAndVersion : ValueSource<TextAndVersion>, ITextVersionable
     {
-        private readonly ITemporaryStorageServiceInternal _storageService;
+        private readonly HostWorkspaceServices _services;
 
-        private SemaphoreSlim? _lazyGate;
-        private ValueSource<TextAndVersion>? _initialSource;
+        // ValueSource<TextAndVersion> or RecoverableText
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs]: i'd doc a little further.  say that it starts as a ValueSource, but can transition to RecoverableText at most once and then latches in that state.

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:48]: there's a race condition in this code (and likely elsewhere).  You could check `if (sourceOrRecoverableText is RecoverableText recoverableText)` and have it be false.  Then, prior to getting to this else block, another thread swaps sourceOrRecoverableText out for a REcoverableText.  Now thsi cast will crash.  

Can we just switch to a nice fat lock (or semaphoreslim)? :)

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs (hunk 1 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:21]: Not easily. Looking into this now.

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:48]: ok.  ignore this comment.  i see that you made a local :)  can you doc that you made a local?

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs [hunk 5]
```diff
@@ -15,129 +16,117 @@ namespace Microsoft.CodeAnalysis
...
             }
 
-            var text = await _text.GetValueAsync(cancellationToken).ConfigureAwait(false);
-            return TextAndVersion.Create(text, _version, _filePath, _loadDiagnostic);
+            var recoverableText = (RecoverableText)_initialSourceOrRecoverableText;
+            return recoverableText.ToTextAndVersion(await recoverableText.GetValueAsync(cancellationToken).ConfigureAwait(false));
         }
 
-        private TextAndVersion InitRecoverable(TextAndVersion textAndVersion)
+        private sealed class RecoverableText : WeaklyCachedRecoverableValueSource<SourceText>, ITextVersionable
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:104]: WeaklyCachedRecoverableValueSource :'(

i will be happy when we can get rid of all these value-sources.

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs (hunk 4 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs]: Done.

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs [hunk 6]
```diff
@@ -15,129 +16,117 @@ namespace Microsoft.CodeAnalysis
...
             {
-                _parent = parent;
+                _storageService = services.GetRequiredService<ITemporaryStorageServiceInternal>();
+
+                Version = textAndVersion.Version;
+                LoadDiagnostic = textAndVersion.LoadDiagnostic;
             }
 
+            public TextAndVersion ToTextAndVersion(SourceText text)
+                => TextAndVersion.Create(text, Version, filePath: null, LoadDiagnostic);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:122]: ok that the filepath is null? seems... concerning.

File: src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs [hunk 7]
```diff
@@ -117,22 +117,11 @@ protected static ValueSource<TextAndVersion> CreateRecoverableText(TextLoader lo
...
-        {
-            get
-            {
-                var recoverableText = this.TextAndVersionSource as RecoverableTextAndVersion;
-                if (recoverableText == null)
-                {
-                    return null;
-                }
-
-                return recoverableText.Storage;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs:133]: note to self.  Have an analyzer that finds thsi pattern and recommends the new form.

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:21]: i think you only use this for GetRequiredService below.  so this should be relaxable.  nbd though.,

**Review by CyrusNajmabadi — APPROVED** [review]
lgtm.  though feeing more into the ValueSource world makes me :'(  will be nice if/when we can burn a lot of this down.

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs (hunk 3 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:85]: I believe so. Interlocked.CompareExchange is the write barrier.

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs (hunk 6 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:122]: the path is unused (obsolete)

**Comment by tmat** [comment]
> though feeing more into the ValueSource world makes me :'( will be nice if/when we can burn a lot of this down.

Yeah, no disagreement here.

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs (hunk 1 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs]: You are right. Done. I was thinking about the caller where the Workspace is still needed.

**Pull request merged**